### PR TITLE
refresh_rhcos.sh: add --raw option

### PR DESCRIPTION
**Pull request by EmilienM**
With -r/--raw, we'll convert the QCOW2 to RAW so the image is stored as
RAW in Glance and Nova won't have to do the conversion each time a new
server is created with that image and that the image isn't in nova
cache.

Signed-off-by: Emilien Macchi <emilien@redhat.com>

### CODE REVIEW

File: refresh_rhcos.sh [hunk 1]
```diff
@@ -8,21 +8,34 @@ if [ -z "$OS_CLOUD" ]; then
 fi
 
 BRANCH="4.2"
+IMAGE_FORMAT="qcow2"
+RAW_CONVERT="False"
```
Comment (mdbooth) [inline refresh_rhcos.sh:12]: You should either use 0 here...

File: refresh_rhcos.sh [hunk 2]
```diff
@@ -8,21 +8,34 @@ if [ -z "$OS_CLOUD" ]; then
...
+        echo "$1 could not be found"
+        exit 1
+    fi
+}
+
 while [[ $# -gt 0 ]]; do
     case "$1" in
+        -r|--raw)
+            check_binary "qemu-img"
+            RAW_CONVERT="1"
```
Comment (mdbooth) [inline refresh_rhcos.sh:32]: ... or True here 😂

File: refresh_rhcos.sh [hunk 3]
```diff
@@ -75,8 +88,16 @@ if ! sha256sum --quiet -c <(echo -n "${IMAGE_SHA}  ${LOCAL_IMAGE_FILE}"); then
     exit 1
 fi
 
+if [ "$RAW_CONVERT" == "1" ]; then
+    IMAGE_FORMAT="raw"
+    IMAGE_BASENAME=$(echo "${LOCAL_IMAGE_FILE}" | cut -f 1 -d '.')
+    echo "Converting QCOW2 image into RAW: ${LOCAL_IMAGE_FILE} to ${IMAGE_BASENAME}.raw with qemu-img"
+    qemu-img convert ${LOCAL_IMAGE_FILE} ${IMAGE_BASENAME}.raw
+    LOCAL_IMAGE_FILE=${IMAGE_BASENAME}.raw
```
Comment (mdbooth) [inline refresh_rhcos.sh:96]: Suggest:

```shell
RAW_IMAGE_FILE=${LOCAL_IMAGE_FILE%.qcow2}.img
echo "Converting QCOW2 image into RAW: ${LOCAL_IMAGE_FILE} to ${RAW_IMAGE_FILE} with qemu-img"
qemu-img convert -f qcow2 "${LOCAL_IMAGE_FILE}" -O raw "${RAW_IMAGE_FILE}"
rm "${LOCAL_IMAGE_FILE}"
LOCAL_IMAGE_FILE="${RAW_IMAGE_FILE}"
```

**Review by mdbooth — COMMENTED** [review]
Note that there's a potential issue with qemu-img convert if we're pulling RHCOS from an untrusted source here: if the RHCOS image specifies a backing file then qemu-img will resolve that backing file locally and upload the contents to glance.

Probably not an issue in context, but I'm not 100% sure what the context is so I'll mention it.

File: refresh_rhcos.sh (hunk 2 above)
Comment (EmilienM) [reply refresh_rhcos.sh:32]: I couldn't make a decision lol

**Comment by mandre** [comment]
I tried something similar in the past with https://github.com/shiftstack/shiftstack-ci/pull/48/ where I detected the glance backend and uploaded the image in the most relevant format. In practice I didn't see improvements on PSI when I tested. Also, the resulting qcow file was really big and I couldn't find how to sparsify it at the time.

**Pull request closed without merging**